### PR TITLE
Restructure Sampler and VentilationStateMachine

### PR DESCRIPTION
**Pull request by avrahamshukron**
# Refactor change
**Functionality is untouched.** The only thing that changed is the structure of the code.
The code is now **50 lines shorter**, and it works **exactly the same** as before, but it is now more maintainable.

# Summary
Responsibility is now self-contained within each class:

## Sampler
Responsible for reading data from sensors and passing it to the state machine.

## VentilationStateMachine
The only stateful class. All responsibilities regarding
	ventilation state and analisys are now contained in this class.

## Benefits
1. Testability - VentilationStateMachine can now be tested for all
corner cases without mocking the world.
2. Readability - The code is much clearer - and even shorter now
3. Robustness - A result of the 2 above.

Signed-off-by: Avi Shukron <avraham.shukron@gmail.com>

### CODE REVIEW

File: tests/unit/test_algo.py [hunk 1]
```diff
@@ -46,11 +48,11 @@ def real_data():
     return t, v
 
 
-def test_slope_recognition(real_data):
+def test_correct_state_transitions(real_data):
     t, v = real_data
-    vmt = VentilationStateMachine({})
-    for ti, vi in zip(t, v):
-        vmt.update(vi, 0, ti)
+    vmt = VentilationStateMachine(Measurements(), Events())
```
Comment (osherdp) [inline tests/unit/test_algo.py:53]: vmt?

File: algo.py [hunk 2]
```diff
@@ -98,7 +89,40 @@ def beat(self, timestamp=None):
         return rate
 
 
+class RunningSlope(object):
+
+    def __init__(self, num_samples=10, period_ms=100):
+        self.period_ms = period_ms
```
Comment (rinslow) [inline algo.py:95]: Unused ?

File: algo.py [hunk 3]
```diff
@@ -98,7 +89,40 @@ def beat(self, timestamp=None):
...
+    Calibration = 0  # State unknown yet
+    Inhale = 1  # Air is flowing to the lungs
+    Hold = 2  # PIP is maintained
+    Exhale = 3  # Pressure is relieved, air flowing out
+    PEEP = 4  # Positive low pressure is maintained until next cycle.
+
+
 class VentilationStateMachine(object):
+
+    NO_BREATH_ALERT_TIME_SECONDS = 12
```
Comment (rinslow) [inline algo.py:125]: get from config

File: algo.py [hunk 4]
```diff
@@ -307,32 +296,11 @@ def read_sensors(self):
...
-        self.peak_flow = max(self.peak_flow, flow_slm)
-
-        if self._config.pressure_range.over(pressure_cmh2o):
-            # Above healthy lungs pressure
-            self._events.alerts_queue.enqueue_alert(AlertCodes.PRESSURE_HIGH)
-        elif self._config.pressure_range.below(pressure_cmh2o):
-            # Below healthy lungs pressure
-            self._events.alerts_queue.enqueue_alert(AlertCodes.PRESSURE_LOW)
+        self.vsm.update(
+            pressure_cmh2o, flow_slm, o2_saturation_percentage, timestamp=ts)
```
Comment (osherdp) [inline algo.py:306]: different order of parameters makes me be scared
can't it be a simple `dataclass` that has those samples?

File: algo.py [hunk 5]
```diff
@@ -121,160 +145,125 @@ class VentilationStateMachine(object):
...
-        self.handlers = state_handlers
+    def __init__(self, measurements, events):
+        self._measurements = measurements
+        self._events = events
         self.log = logging.getLogger(self.__class__.__name__)
         self.rs = RunningSlope(num_samples=7)
-
+        self._config = Configurations.instance()
+        self.last_breath_timestamp = time.time()
         self.current_state = VentilationState.PEEP
```
Comment (rinslow) [inline algo.py:155]: Calibration ?

File: tests/unit/test_algo.py (hunk 1 above)
Comment (avrahamshukron) [reply tests/unit/test_algo.py:53]: oops. `vsm`. typo

File: algo.py (hunk 4 above)
Comment (avrahamshukron) [reply algo.py:306]: And then you will be scared about the order of parameters when initializing the data class...
I'll switch to kwargs.

File: algo.py [hunk 6]
```diff
@@ -45,14 +44,6 @@ def process(self, pressure):
...
 
-class VentilationState(Enum):
-    Calibration = 0  # State unknown yet
-    Inhale = 1  # Air is flowing to the lungs
-    Hold = 2  # PIP is maintained
-    Exhale = 3  # Pressure is relieved, air flowing out
-    PEEP = 4  # Positive low pressure is maintained until next cycle.
-
-
 class RateMeter(object):
```
Comment (rinslow) [inline algo.py:47]: I know it's now part of this issue, but I'm reviewing this file in it's entirety, 
This class is not very readable nor it is easy to understand how to use it and what it does.
From what I see it keeps samples in a `deque` for a certain span of time right? 

That's kind of hard to understand from just the names `RateMeter` or `beat()` 

Either rename to make it self-explanatory or document and explain

File: algo.py [hunk 7]
```diff
@@ -98,7 +89,40 @@ def beat(self, timestamp=None):
...
+        self.period_ms = period_ms
+        self.max_samples = num_samples
+        self.data = deque(maxlen=num_samples)
+        self.ts = deque(maxlen=num_samples)
+
+    def reset(self):
+        self.data.clear()
+        self.ts.clear()
+
+    def add_sample(self, value, timestamp=None):
```
Comment (rinslow) [inline algo.py:104]: Should be splitted to `add_sample()` and `slope()`, this is kind of un-natural to see 

```python
slope = self.running_slope.add_sample(Sample(...))
```

But rather 
```python
self.running_slope.add_sample(Sample(...))
new_slope = self.running_slope.calculate_slope()
```

File: algo.py [hunk 8]
```diff
@@ -121,160 +145,125 @@ class VentilationStateMachine(object):
         VentilationState.Exhale: (float.__gt__, EXHALE_TO_PEEP_SLOPE, VentilationState.PEEP),
```
Comment (rinslow) [inline algo.py:145]: Maybe do in a future issue, but this tuple of a function, and 2 states isn't really readable, 
why not wrap it up nicely in a `Transition` class?

File: algo.py [hunk 9]
```diff
@@ -121,160 +145,125 @@ class VentilationStateMachine(object):
...
+        self.min_pressure = min(self.min_pressure, pressure_cmh2o)
+        self.peak_flow = max(self.peak_flow, flow_slm)
+
+        if self._config.pressure_range.over(pressure_cmh2o):
+            # Above healthy lungs pressure
+            self._events.alerts_queue.enqueue_alert(AlertCodes.PRESSURE_HIGH)
+        elif self._config.pressure_range.below(pressure_cmh2o):
+            # Below healthy lungs pressure
+            self._events.alerts_queue.enqueue_alert(AlertCodes.PRESSURE_LOW)
+
```
Comment (rinslow) [inline algo.py:233]: There should be two functions here
`update_measurements()` and `publish_alerts()`

File: algo.py [hunk 10]
```diff
@@ -121,160 +145,125 @@ class VentilationStateMachine(object):
...
+        self.check_transition(pressure_cmh2o, timestamp)
+
+    def check_transition(self, pressure_cmh2o, timestamp):
```
Comment (rinslow) [inline algo.py:236]: It's really not checking a transition I think it calculates the next state

File: algo.py [hunk 11]
```diff
@@ -121,160 +145,125 @@ class VentilationStateMachine(object):
...
+        # Now, calculate the current slope of the pressure graph to see in which
+        # ventilation state we are.
+        slope = self.rs.add_sample(pressure_cmh2o, timestamp)
+        if slope is None:
+            return  # Not enough data
+
+        func, threshold, next_state = self.TRANSITION_TABLE.get(self.current_state)
+        if func(slope, threshold):
+            self.rs.reset()  # TODO: Rethink. Maybe not reset?
```
Comment (rinslow) [inline algo.py:245]: Can you explain what benefit resetting the running slope gives? 
Does it produce better accuracy? can we test it?

File: algo.py [hunk 12]
```diff
@@ -98,7 +89,40 @@ def beat(self, timestamp=None):
...
+
+class VentilationState(Enum):
+    Calibration = 0  # State unknown yet
+    Inhale = 1  # Air is flowing to the lungs
+    Hold = 2  # PIP is maintained
+    Exhale = 3  # Pressure is relieved, air flowing out
+    PEEP = 4  # Positive low pressure is maintained until next cycle.
+
+
 class VentilationStateMachine(object):
```
Comment (rinslow) [inline algo.py:123]: I really really love the thinking and algorithmics here but it's really not a state machine 
it 
* calculates the next state based on given input
* publishes measurement results
* publishes alerts 

It's really a cluster of three classes made into one big god algo-object

File: algo.py (hunk 4 above)
Comment (osherdp) [reply algo.py]: yes, keyword arguments were what I meant

**Pull request merged**